### PR TITLE
Replace optparse with argparse

### LINKED ISSUE #1778: Switch DbCLI scripts from optparse to argparse
The scripts in `$RDBASE/Projects/DbCLI` currently use Python's old optparse package to handle command line arguments. They should be migrated to use the new argparse package (https://docs.python.org/3/library/argparse.html).

### PR DESCRIPTION
**Pull request by gedeck**
#### Reference Issue
Fixes #1778

#### What does this implement/fix? Explain your changes.
Replaces optparse with argparse. I highlight specific changes, when they are not obvious due to reformatting.

#### Any other comments?
- Changes are minimal
- Could organise the arguments into groups
- there is some duplication in the command line arguments of the two scripts

### CODE REVIEW

File: Projects/DbCLI/CreateDb.py [hunk 1]
```diff
@@ -32,9 +32,7 @@
 # Created by Greg Landrum, July 2007
 _version = "0.13.0"
 
-_usage = """
- CreateDb [optional arguments] <filename>
```
Comment (gedeck) [inline Projects/DbCLI/CreateDb.py]: This is automatically added argparse

File: Projects/DbCLI/CreateDb.py [hunk 2]
```diff
@@ -69,85 +69,94 @@
...
-                  help='the column index with mol names')
+
+def initParser():
+  """ Initialize the command line parser """ 
+  parser = argparse.ArgumentParser(usage='CreateDB [optional arguments] <filename>',
+                                   description=_description,
+                                   formatter_class=argparse.RawDescriptionHelpFormatter)
+
+  parser.add_argument('filename', nargs='?', help='File containg molecules to load into database')
+  parser.add_argument('--version', action='version', version='%(prog)s ' + _version)
```
Comment (gedeck) [inline Projects/DbCLI/CreateDb.py]: - positional arguments must be explicitly defined
- `--version` returns version number

File: Projects/DbCLI/CreateDb.py [hunk 3]
```diff
@@ -69,85 +69,94 @@
...
+                    help='do not provide status messages')
+  
+  parser.add_argument('--molFormat', default='', choices=('smiles', 'sdf', ''),
+                    help='specify the format of the input file')
+  parser.add_argument(
+    '--nameProp', default='_Name',
+    help='specify the SD property to be used for the molecule names. Default is to use the mol block name')
+  parser.add_argument(
+    '--missingPropertyVal', default='N/A',
+    help='value to insert in the database if a property value is missing. Default is %(default)s.')
```
Comment (gedeck) [inline Projects/DbCLI/CreateDb.py]: Here is another change `%(default)s`

File: Projects/DbCLI/CreateDb.py [hunk 4]
```diff
@@ -69,85 +69,94 @@
...
+  
+  parser.add_argument('--noMorganFps', '--noCircularFps', default=True, dest='doMorganFps',
+                    action='store_false', help='skip calculating Morgan (circular) fingerprints')
+  parser.add_argument('--morganFpTableName', default='morganfps',
+                    help='name of the Morgan fingerprints table')
+  
+  parser.add_argument('--delimiter', '--delim', default=' ', help='the delimiter in the input file')
+  parser.add_argument('--titleLine', default=False, action='store_true',
+                    help='the input file contains a title line')
+  parser.add_argument('--smilesColumn', '--smilesCol', default=0, type=int,
```
Comment (gedeck) [inline Projects/DbCLI/CreateDb.py]: And the type is no longer defined as a string, but as a function

File: Projects/DbCLI/CreateDb.py [hunk 5]
```diff
@@ -431,11 +440,12 @@ def CreateDb(options, dataFilename='', supplier=None):
 
 
 if __name__ == '__main__':
-  options, args = parser.parse_args()
+  parser = initParser()
+  options = parser.parse_args()
   if options.loadMols:
-    if len(args) != 1:
+    if options.filename is None:
```
Comment (gedeck) [inline Projects/DbCLI/CreateDb.py]: We don't need to check the length here, as filename can only occur 0 or 1 times

File: Projects/DbCLI/SearchDb.py [hunk 6]
```diff
@@ -509,116 +511,122 @@ def poolFromCurs(curs, similarityMethod):
...
+  options = parser.parse_args()
+  
+  if options.filename is None and not (options.smilesQuery or options.smartsQuery or options.propQuery):
     parser.error('please either provide a query filename argument or do a data or smarts query')
 
-  if len(args):
-    queryFilename = args[0]
-  else:
-    queryFilename = None
+  queryFilename = options.filename
```
Comment (gedeck) [inline Projects/DbCLI/SearchDb.py]: No need to use the if-statement. If filename isn't given on the command line, `options.filename` is `None`

File: Projects/DbCLI/SearchDb.py [hunk 7]
```diff
@@ -509,116 +511,122 @@ def poolFromCurs(curs, similarityMethod):
...
+  parser = argparse.ArgumentParser(usage='SearchDB [optional arguments] <sdffilename>',
+                                   description=_description,
+                                   formatter_class=argparse.RawDescriptionHelpFormatter)
+  
+  parser.add_argument('filename', nargs='?', help='File containg molecules for searching')
+  parser.add_argument('--version', action='version', version='%(prog)s ' + _version)
+
+  parser.add_argument(
+    '--dbDir', default='/db/camm/CURRENT/rdk_db',
+    help='name of the directory containing the database information. The default is %(default)s')
```
Comment (gedeck) [inline Projects/DbCLI/SearchDb.py]: I think we should replace the default here with the same as in the CreateDB script:
```
 parser.add_argument('--dbDir', default='', help='name of the directory containing the database information')
```

Comment (greglandrum) [reply Projects/DbCLI/SearchDb.py]: Agreed. That old reference to `/db/camm` should be removed

File: Projects/DbCLI/SearchDb.py [hunk 8]
```diff
@@ -509,116 +511,121 @@ def poolFromCurs(curs, similarityMethod):
...
+  """ Initialize the command line parser """ 
+  parser = argparse.ArgumentParser(usage='SearchDB [optional arguments] <sdffilename>',
+                                   description=_description,
+                                   formatter_class=argparse.RawDescriptionHelpFormatter)
+  
+  parser.add_argument('filename', nargs='?', help='File containg molecules for searching')
+  parser.add_argument('--version', action='version', version='%(prog)s ' + _version)
+
+  parser.add_argument('--dbDir', default='',
+    help='name of the directory containing the database information. The default is the current directory')
```
Comment (gedeck) [inline Projects/DbCLI/SearchDb.py]: Removed the Novartis directory reference

File: Projects/DbCLI/SearchDb.py [hunk 9]
```diff
@@ -124,8 +126,8 @@ def GetMolsFromSmilesFile(dataFilename, errFile, nameProp):
       continue
     m = Chem.MolFromSmiles(smi)
     if not m:
-      if errfile:
-        print(idx, nm, smi, file=errfile)
+      if errFile:
+        print(idx, nm, smi, file=errFile)
```
Comment (gedeck) [inline Projects/DbCLI/SearchDb.py]: The function argument is `errFile`.

**Review by gedeck — COMMENTED** [review]
This should be ready to go now

**Comment by greglandrum** [comment]
Thanks Peter

**Pull request merged**